source/transfer_list.rst

```diff
@@ -221,24 +230,19 @@ The intent of the void entry type is for information to be removed from the list
      - 0x0
      - The tag_id field must be set to **0**.
 
-   * - hdr_size
-     - 0x4
-     - 0x4
-     - The size of this entry header in bytes.
-
    * - data_size
+     - 0x3
      - 0x4
-     - 0x8
-     - The size of the data content in bytes.
+     - The size of the void space in bytes.
 
-   * - reserved
-     - 0x4
-     - 0xc
-     - Reserved, must be zero.
+   * - hdr_size
+     - 0x1
+     - 0x7
+     - The size of this entry header in bytes must be set to **8**.
```

Comment:
It would be good to not explicitly write "8" everywhere to ease maintenance. Why don't we define a constant in this doc (e.g. ENTRY_HDR_SIZE) and use that throughout? Also, it would be good to clarify that TL producers complying to this version of the spec must set this field to 8, although consumers of the TL must be robust to higher values.

Reply:
Do you mean a constant inside the .rst document that would then be displayed as 8 when the document is rendered (does .rst support such a thing), or should it say ENTRY_HDR_SIZE in the final text? I think the spec would be easier to read if it just showed the correct number directly. (In general it might be nice to deduplicate some of this stuff in the .rst source, e.g. the parts of the header that are always the same... I'm not sure if that's easily possible with .rst.)

Your version comment brings up an important point that isn't really well defined in the spec yet, either: what does it mean for a reader to encounter a `version` field in the TL header that's newer than what it knows? Should it reject the whole list or should it continue parsing it as if it were the older version? And what if it's a writer instead of a reader -- is it legal to append tags to an existing list of a newer version? Those things should be clearly written down somewhere.

Reply:
> Do you mean a constant inside the .rst document that would then be displayed as 8 when the document is rendered (does .rst support such a thing)

I hadn't thought of that but it seems rst does allow this:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions

As you say, I keen to avoid duplicating this stuff in the source to reduce the maintenance burden.

>  what does it mean for a reader to encounter a version field in the TL header that's newer than what it knows? 

I think readers should continue parsing as if it were the old version (at least that's what I'd assumed). Also if it encounters older versions, it should be able to parse those older versions. A writer can append tags to a list of a newer version, but must conform to the version spec of the spec it knows about. I agree this needs writing down (maybe in a separate commit/PR).

This also opens the question of whether we support major/minor versions, where differences in the major version mean the above compatibility guarantees don't apply. But that's hard to do in a 1 byte version field. Maybe it's not worth packing the TL header down to 16 bytes after all?

Reply:
Thanks, I will look into the substitution feature to try to reduce some of the duplication.

I think a major version change (i.e. one where there are no compatibility guarantees at all and an outdated reader should just give up) would be best represented by changing the signature field, because it's basically a different spec at that point?

Reply:
After playing around with the GitHub .rst implementation for a bit I unfortunately couldn't find any good way to do this. It seems that the substitution replace:: feature can only really insert raw text and not any structure elements like table rows and columns. There seems to be a more powerful directive called `.. include`, but unfortunately that one is disabled in the GitHub engine which makes it not a good fit for our purposes.

What do you think about just using commented C headers instead of .rst tables to define data layouts for individual TE types (with a few rules, e.g. that all struct members must be naturally aligned and any reserved/padding space must be explicitly created as a member field)? I think that format would be more easy to write / maintain for most contributors anyway, and it could conveniently serve as a starting point for implementations as well.

Reply:
> I think a major version change (i.e. one where there are no compatibility guarantees at all and an outdated reader should just give up) would be best represented by changing the signature field, because it's basically a different spec at that point?

That's a good idea - works for me.

> It seems that the substitution replace:: feature can only really insert raw text and not any structure elements like table rows and columns.

That's a shame.

> What do you think about just using commented C headers instead of .rst tables to define data layouts for individual TE types

I don't have any objections to this but I'm not sure how that would reduce duplication (e.g. wouldn't you still need insert "8" in the C comment of all TE types?).

Reply:
...but how about (in a separate PR) we add C code into the same repo in separate files? I can adjust the U-Boot impl and add impl, header and test files

Reply:
Technically that sounds OK but it would change the nature of the repo. This is currently Creative Commons licensed and if we add actual C files (as oppose to code fragments in docs), they'll need a different license (e.g. BSD-3-Clause). At least on the Arm side there will be more paperwork to do so I'm not that keen unless there's a clear win.

Reply:
My point was that with C code you can do something like this:
```
struct te_fdt {
  struct te_header h;
  uint8_t fdt[];  /* The FDT blob. */
};
```
So you can just reference the TE header structure in each individual TE structure and don't have to keep copying the header fields everywhere. I was trying to find a solution to deduplicate the whole header, not just the number 8 (that one you could actually do with the substitution feature).

Reply:
But that doesn't need C. We could replace the repeating header with one row which says there is a header. with tag = 2, etc. The header size is the same for most entries so doesn't need to be mentioned.

But I suggest we worry about this in another PR. In fact even this PR seems to have too much going on. We can't easily submit individual commits in the (terrible) github review tool.

Reply:
Well, I can't split this into mutliple PRs because it's all touching the same tables in the same file, so the patches depend on each other. (Why are we hosting this on GitHub again if we all agree that it's terrible to use? We would have had a nice Gerrit on trustedfirmware.org...)

Reply:
I know, I know :-) It is possible to set up gerrit for use with github?

Reply:
IIRC best we could come up with when we explored this for TF-A itself was to host it externally with Gerrit and set GitHub up to be a read-only mirror of that repo (like https://github.com/ARM-software/arm-trusted-firmware is for TF-A right now). New patches need to go through the Gerrit. (TF-A Gerrit supports SSO with GitHub IDs, so getting set up there is easy and painless for existing GitHub users.)

Reply:
I now used a .rst substitution to substitute the whole description line for hdr_size in the latest version, which I think is the  best we can do with .rst. I still think just writing C structures would make this easier to read and maintain. I'm not sure just putting a "header comes here" line in the table would look good, but we can experiment with that in a later PR.

Reply:
> IIRC best we could come up with when we explored this for TF-A itself was to host it externally with Gerrit and set GitHub up to be a read-only mirror of that repo

That was the best we came up with but there are potentially better solutions. I know there are tools that automatically sync GitHub PRs and Gerrit reviews and there's a long-standing task to investigate this more (stalled on more important OpenCI work):
https://developer.trustedfirmware.org/T856

Also, using Gerrit means self-hosting somewhere and TF.org was considered not neutral enough. Anyway, this is off-topic.

> we can experiment with that [using C structures] in a later PR.

Agreed.